### PR TITLE
Undersampling project proposal

### PR DESCRIPTION
**Pull request by kindalime**
Proposal for the undersampling project discussed in [#33](https://github.com/tensorflow/tfx-addons/issues/33).

### CODE REVIEW

**Comment by kindalime** [comment]
@googlebot I signed it!

**Comment by kindalime** [comment]
@googlebot I signed it!

**Review by rcrowe-google — COMMENTED** [review]
This looks good overall!  I would discourage the idea of using BigQuery for this, since that would limit it to use on the Google Cloud only.

**Comment by rcrowe-google** [comment]
There is a [related discussion and implementation in the TFX repo](https://github.com/tensorflow/tfx/issues/3831#issuecomment-857432043).

**Comment by kindalime** [comment]
Hello! Just a quick update: I've completed an initial version of the undersampling component, which can be found [here](https://github.com/kindalime/tfx-addons/tree/sampling/projects/sampling). Constructive criticism is much appreciated!

File: proposals/undersampling_project_proposal.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+**Your email:** danielk@twitter.com
+
+**Your company/organization:** Twitter
+
+**Project name:** Undersampling Component
+
+## Project Description
+
+This project will be a custom function-based component that inputs an artifact in `tfRecord` format of `tf.Example`s and randomly undersamples it, reducing the data to the lowest-frequency class. It will primarily use an underlying Apache Beam pipeline that will be wrapped inside the TensorFlow component.
```
Comment (zhitaoli) [inline proposals/undersampling_project_proposal.md:17]: (non blocking)

How much do you imagine this implementation will be tied to either the format or the tf.Example payload? If it becomes a different container, or a different payload, would sample still work?

File: proposals/undersampling_project_proposal.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Project Implementation
+
+At a high level, the plan is to use Apache Beam to injest a `tfRecord` of `tf.Examples`, shuffle them, convert them into a key-value `PCollection` with keys as class values and values as data points, and then use `Sample.FixedSizePerKey()` in order to perform the actual undersampling. The algorithm will be written as an Apache Beam pipeline, which will be wrapped into a TensorFlow custom function component to use with TFX pipelines. The component would be written as inputting a `tfRecord` artifact and exporting a similar `tf.Record` artifact, making its placement in a pipeline nearly ubiquitous. If necessary, the component may also be changed to a fully-custom component, albeit one where only the executor is edited.
```
Comment (zhitaoli) [inline proposals/undersampling_project_proposal.md:35]: ingest

Comment (zhitaoli) [reply proposals/undersampling_project_proposal.md:35]: What's the use case of TensorFlow custom function, vs just another Python function?

File: proposals/undersampling_project_proposal.md (hunk 1 above)
Comment (kindalime) [reply proposals/undersampling_project_proposal.md:17]: While this implementation is tied to the tf.Example payload, this is the standard pipeline data flow for pipelines that are downstream from ExampleGen. We are planning for data format logic to be handled by ExampleGen; as long as your ExampleGen component or a custom component can ingest a requested data format, the Undersampler can use it.

File: proposals/undersampling_project_proposal.md (hunk 2 above)
Comment (kindalime) [reply proposals/undersampling_project_proposal.md:35]: This is actually outdated, but the use of a TFX custom component function would be to allow its use in TFX pipelines as a component. The current implementation uses a fully-custom component with a custom executor and spec, which fulfills a similar role while allowing more complexity.

**Comment by kindalime** [comment]
Just a quick update: a finished initial version of the component can be found [here](https://github.com/tensorflow/tfx-addons/pull/37).

File: pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
+
+# Background
+
+# Problem
+
+# Solution
+
+Fixes # (issue)
```
Comment (casassg) [inline pull_request_template.md:9]: Can we remove this from here.

**Review by casassg — APPROVED** [review]

File: pull_request_template.md (hunk 3 above)
Comment (kindalime) [reply pull_request_template.md]: Good catch - just removed it.

**Comment by rcrowe-google** [comment]
Closed accidentally

**Pull request closed without merging**